unittests/test_pipeline.py

```diff
@@ -307,6 +311,94 @@ def test_extra_resources(self):
         self.assertCountEqual(expected_job_options, test.job.options)
 
 
+class TestNewStyleChecks(unittest.TestCase):
+    def test_regression_test(self):
+        class MyTest(RegressionTest):
+            def __init__(self, a, b):
+                super().__init__()
+                self.a = a
+                self.b = b
+
+        test = MyTest(1, 2)
+        self.assertEqual(os.path.abspath(os.path.dirname(__file__)),
+                         test.prefix)
+        self.assertEqual('my_test_1_2', test.name)
+
+    def test_regression_test_strange_names(self):
+        class C:
+            def __init__(self, a):
+                self.a = a
+
+            def __repr__(self):
+                return 'C(%s)' % self.a
+
+        class MyTest(RegressionTest):
+            def __init__(self, a, b):
+                super().__init__()
+                self.a = a
+                self.b = b
+
+        test = MyTest('(a*b+c)/12', C(33))
+        self.assertEqual('my_test__a_b_c__12_C_33_', test.name)
+
+    def test_user_inheritance(self):
+        class MyBaseTest(RegressionTest):
+            def __init__(self, a, b):
+                super().__init__()
+                self.a = a
+                self.b = b
+
+        class MyTest(MyBaseTest):
+            def __init__(self):
+                super().__init__(1, 2)
+
+        test = MyTest()
+        self.assertEqual('my_test', test.name)
```

Comment:
As we enforce since version 2.12 that the names of the loaded tests must be unique (https://eth-cscs.github.io/reframe/running.html?highlight=unique#discovery-of-regression-tests), I have the following doubt. Don't we get non-unique names e.g. in the following scenario?
```
class BaseMatrixVectorTest(RegressionTest):
    def __init__(self):
    ...
...

class SerialTest(BaseMatrixVectorTest):
    def __init__(self):
    ...
...
```

...and (in another file):

```
class BaseStencilTest(RegressionTest):
    def __init__(self):
    ...
...

class SerialTest(BaseStencilTest):
    def __init__(self):
    ...
...
```

We get in both cases `test.name` equal to `serial_test`, right? If I am not overseeing something and it is like that, then I think this must be avoided, because it is not an unlikely scenario. One solution could be to prefix the generated names additionally with the name(s) of the super class(es).

Reply:
Personally, I can't see this as a blocking issue for this PR. The class name is a guess and a convenience; the users may always set their own name if a conflict occurs or if they don't trust the auto-generation mechanism. Whichever heuristic we had with the name, there would always be a case to fool it. In your scenario, it could be easily fooled if the same class names were also present at another file. Then, we end up having to include the name of the file. Oops, this is not enough either! Because, we might have two files with the same class structure in different directories and so on an so forth. We would end with too long, unusable test names, forcing the users to use their own...  After all, the conflict detection mechanism is still valid, since it checks the names during loading the test.

Reply:
I totally agree that the test name generation does not need to be bullet proof; it only needs to catch the common scenarios (as we have discussed already some days ago off-line). Now, how common the above scenario is depends very much on the programming habits of the users.

Besides the uniqueness issue, the test name should be most meaningful but not too long as you pointed out. One could maybe use somehow the name of the direct superclass (if it is defined by the user) to make the test names more meaningful. However, this could already make the name too long...

In any case this issues are certainly not blocking as you say. You may merge it as is and we may see later if the automatic test name generation can still be improved...

Reply:
To fix this, I'm gonna use the fully qualified name of the class from the `__qualname__` attribute.

Reply:
@omlins Prefixing with the module name (or anything derived from the filesystem) is not a good idea. The `__qualname__` does not do that deliberately. Have a look [here](https://www.python.org/dev/peps/pep-3155/#excluding-the-module-name). I suggest getting along with just the qualified name (which covers also nested classes compared to `__name__`) and if a conflict occurs, then the user may define its own name.